Typography component

daksas303 opened the pull request:
Added Typography component
Typography have these options:

- Element type (`h1`, `h2` ..., `p`), all element types have already the default size for mobile and desktop according to Figma https://www.figma.com/file/ctDKMhAPIjLUVNNmq430D4/DS-ESBS%3A-Foundations?node-id=73%3A133&mode=dev
- Element size. same options as element type, plus sizes `h1-hero`, `h1-form`, `p-small`, and `p-large`. Element size can be set either as single string value `size="h1-hero"` as such both mobile and desktop sizes will be the same, or you can pass array of exactly two values, first for mobile and second for desktop `size={['h1', 'h1-hero]}`
- Element variant options define font-weight, `light`, `normal`, `medium`, and `bold`. Default values are `bold` for headers and `normal` for paragraph

For normalizing text we used utility from mestskakniznica.
Also for merging tailwind classes, we are using twMerge, which ensures the precedence of tailwind classes that are passed from the user.

mpinter commented on src/showcases/TypographyShowcase.tsx:
```diff
@@ -0,0 +1,132 @@
+
+const TypographyShowcase: React.FC = () => (
+  <>
+    <div className="h-[100%] border-b-2 border-indigo-500">
```
there's a lot of repetition in these divs, can it be a reusable component ? do we need the h-[100%] everywhere ?

daksas303 replied:
Made as a reusable component, also `h-[100%]` and other unnecessary styles were removed.

mpinter commented on tailwind.config.js:
```diff
@@ -1,4 +1,9 @@
 /** @type {import('tailwindcss').Config} */
+const defaultTheme = require('tailwindcss/defaultTheme')
+
+const toRem = (px) => `${px / 16}rem`
+const getFontSize = (size) => [toRem(size[0]), toRem(size[1])]
```
I'm not sure if this is a good idea, but open to discussion. If it's form your head it deserves an explanation in comment. If it's taken form somewhere it deserves linking to source

daksas303 replied:
This is taken from bratislava.sk tailwind config. Generally, I think the usage of rems adds a bit of clarity and removes a lot of arbitrary values, but I would say that writing the size in pixels and then just converting it possibly loses this point. I was just following the general flow of Bratislava, and here I would be fine with both options tho thinking about it now I would probably lean more towards removing it. If you agree we can get rid of this.

https://github.com/bratislava/bratislava.sk/blob/master/next/tailwind.config.js#L10

mpinter commented on tailwind.config.js:
```diff
@@ -8,6 +13,42 @@ export default {
+    },
+
+    fontSize: {
+      'bcl-h1-hero': getFontSize([56, 64]),
```
what is bcl ? is it a custom prefix for our classes ? document if so

daksas303 replied:
Exactly, there was a conflict with some custom classes in konto. Added commentary in code.

mpinter reviewed (changes requested):
- in general I'm missing a lot of documentation/commentary - what is in PR conversation in here is documentation that belongs to the codebase, along with a few more mentioned in comments below. I'd keep most documentation next to the code as comments.
- I see many more options of text styling available than we want to allow - the goal here is to restrict rather then make it configurable. I.e. headers shouldn't have variants for boldness. I think the only concession we want to make is the ability to separate semantics and styling - that is, to have a h2 semantically but for it to be in the styling of h3. 
- other than that, this should allow no more options that you can see in Figma - 7 types of headers (h1-hero to h6) with their mobile versions, without any further variations available + 2 'button' texts + 3 * 4 paragraph variants. So, the 'default' should only be 21 'configurations' altogether of Typography. By the way, the typography figma should be linked in the code as well.
- if there are any particular reasons why we need further customisation or other variants, there should be reasoning for it in code

Following up on all that - I'm not sure if the size + variant approach is correct here. But I'd come to that and the review of component itself after we address the above.

daksas303 commented:
> * in general I'm missing a lot of documentation/commentary - what is in PR conversation in here is documentation that belongs to the codebase, along with a few more mentioned in comments below. I'd keep most documentation next to the code as comments.
> * I see many more options of text styling available than we want to allow - the goal here is to restrict rather then make it configurable. I.e. headers shouldn't have variants for boldness. I think the only concession we want to make is the ability to separate semantics and styling - that is, to have a h2 semantically but for it to be in the styling of h3.
> * other than that, this should allow no more options that you can see in Figma - 7 types of headers (h1-hero to h6) with their mobile versions, without any further variations available + 2 'button' texts + 3 * 4 paragraph variants. So, the 'default' should only be 21 'configurations' altogether of Typography. By the way, the typography figma should be linked in the code as well.
> * if there are any particular reasons why we need further customisation or other variants, there should be reasoning for it in code
> 
> Following up on all that - I'm not sure if the size + variant approach is correct here. But I'd come to that and the review of component itself after we address the above.

I added commentary to each showcase component about usage and also commented out some possibly unclear types, used utils functions, added figma link, commentary about used prefixes, etc.

Also,  I restricted variants via TS to by default allow just the provided options from Figma, so when the user uses for example `type="h1"` it is now NOT possible to use at the same time variant="light" to change the default `font-weight: bold`.  I tried to explain in comments the usage for each type `header`, `p`, and `span`. 

Just to clarify I believe there are actually 22 configurations, because of the `H1 form header` variant. They are currently all shown on our local showcase page. 

Changed the component structure a bit, to make it a bit less overwhelming and more organized.

mpinter commented:
@daksas303 thanks :+1: - I get the idea now, will let @radoslavzeman and @MarekBodingerBA go through this. When everyone is down with the "api" of the component we might do some more finetuning.

radoslavzeman commented on tailwind.config.js:
```diff
@@ -8,6 +13,48 @@ export default {
+      'bcl-h2': getFontSize([32, 40]),
+      'bcl-h3': getFontSize([28, 36]),
+      'bcl-h4': getFontSize([24, 32]),
+      'bcl-h4-mobile': getFontSize([18, 26]),
```
`bcl-h4-mobile` ?

radoslavzeman reviewed (commented):
Please change explanation comments to TS doc https://tsdoc.org/.

daksas303 commented on tailwind.config.js:
```diff
@@ -8,6 +13,48 @@ export default {
+      light: '300',
+      normal: '400',
+      medium: '500',
+      bold: '600',
```
`bold` renamed to `semibold`

daksas303 commented on tailwind.config.js:
```diff
@@ -8,6 +13,48 @@ export default {
+      md: '768px',
+      lg: '1216px',
+      xl: '1280px',
+    },
```
`Screens` removed, so now we are using default tailwind screen sizes => https://tailwindcss.com/docs/responsive-design

If we decide in the future that we want to use our own values different from these, we will adjust accordingly.

daksas303 commented on package.json:
```diff
@@ -71,6 +72,7 @@
     "react-aria-components": "^1.0.0-beta.2",
     "react-dom": "^18.2.0",
     "tailwindcss": "^3.3.5",
+    "tailwind-merge": "^2.0.0",
```
Removed `tailwind-merge` from dependencies

daksas303 commented on lib/components/Typography.tsx:
```diff
@@ -0,0 +1,66 @@
+import { typographyElementDefaultStyles, typographyFontWeightVariants } from '../utils/constants'
+import { getCustomSize } from '../utils/getCustomSize'
+
+type SupportedVariantType = 'light' | 'normal' | 'medium' | 'bold'
```
Changed `variant` prop to `fontWeight` prop. Supported options for `fontWeight` are now `light`, `normal`, `medium` and `semibold`.

daksas303 commented on lib/utils/constants.ts:
```diff
@@ -0,0 +1,40 @@
+  'p-small': 'text-bcl-p-sm font-normal lg:text-bcl-p-sm',
+  'p-large': 'text-bcl-p-md font-normal lg:text-bcl-p-md ',
+  span: 'text-bcl-span-base font-bold lg:text-bcl-span-base',
+  'span-large': 'text-bcl-span-large font-bold lg:text-bcl-span-large',
```
The span size is set in tailwind.config.js

```
 'bcl-span-base': getFontSize([16, 24]),
 'bcl-span-large': getFontSize([20, 32]),
```

If you mean where I took the values, its from figma => https://www.figma.com/file/ctDKMhAPIjLUVNNmq430D4/DS-ESBS%3A-Foundations?type=design&node-id=73-126&mode=design&t=0nJLIrLCtwQHsm7k-0

`Large button` and `Default button`

radoslavzeman replied:
Hmm this seems a bit confusing that "span" sizes are used for buttons. I understand that we need to apply styles to some html element, and span is the one that fits for this case.
But maybe we don't have to sue Typography for Buttons at all and style Buttons separately without Typography (as it is done now).

I don't want to block this PR because of this, so for me, adding a TODO comment about "span used for buttons" would be enough for now. @mpinter @MarekBodingerBA ?

daksas303 replied:
I understand it's not ideal to use `type="span"` in our case where we expect to use it for buttons. But I was thinking about this type to be perhaps more universal, for example, that it could be used also in the one-line variant of the Alert component.

I agree that we could just add a TODO comment at this point and get back to it when we figure out what to do next with this.

daksas303 replied:
Added TODO comment.

mpinter replied:
I don't have a strong opinion here, I'm fine with the 'status quo' of this being span - so ok with merging this with a todo

mpinter reviewed (commented):
Current version makes sense to me. One thing I'd like to clean up with design are the (optional) responsive paragraphs. From discussions with Patrik (our designer) it sounds like the responsiveness is intended to always behave the same way - responsive for large paragraphs, non-responsive form other variants. If that's the case we may get rid of this param and enforce this strategy.

Once we sync on that and if the others are fine with current api, it's good to from my POV.

daksas303 commented on lib/components/Typography.tsx:
```diff
@@ -0,0 +1,94 @@
+  responsive?: T extends 'p' ? boolean : never
+}
+
+// eslint-disable-next-line no-secrets/no-secrets
```
I turned the rule off and removed the disable next line. Wanted to remove this particular plugin altogether, but it seems `eslint-config-auto` relies on it as a dependency and results in an error when removing `eslint-plugin-no-secrets` making eslint not work at all. 

So let's wait for the final decision regarding the paragraphs and I believe after figuring that out we are good to go.

radoslavzeman replied:
Yes, in our setup using `eslint-config-auto`, the best approach is just to add `'no-secrets/no-secrets': 'off'` to eslintrc (as it is done now). 👍

radoslavzeman replied:
I'll discuss paragraph sizing with Patrik today and will let you know.

radoslavzeman replied:
Paragraph sizing confirmed, as mentioned here before:
large - zmensi sa na medium
medium - zostava rovnaky
small - zostava rovnaky

MarekBodingerBA commented on lib/components/Typography.tsx:
```diff
@@ -0,0 +1,95 @@
+  fontWeight,
+  ...otherAttributes
+}: TypographyProps<T>) => {
+  const CustomTag = type as SupportedElementType
```
Is this casting needed?

radoslavzeman replied:
Please be consistent in naming: `CustomTag` --> `CustomElement`

daksas303 replied:
Changed the naming.

As for the casting, it is somehow needed, because `type` is of Type `T extends SupportedElementType` and at that point when I try to use the type as `CustomElement` the TS throws an error.

mpinter commented on lib/utils/constants.ts:
:+1: agreed (does not have to be part of this PR, but let's start doing that)

daksas303 replied:
Alright, I also think this approach will be more clear. When the PR gets approved I am going to squash the commits, so I will include this changed structure before merging.

radoslavzeman reviewed (approved):
Okey, it's in good shape for me to merge.

I would like to see in further PR:
- components to have parent folder and all related files in that folder
- path aliases, instead of relative paths
- styleguide helpser compoents same as in konto (Wrapper, Stack)
- showcase can be more condensed, headings can be in one stack for example
- more consistency in naming
- less or better none type casting (`as SomeType`)

radoslavzeman commented on lib/utils/constants.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+// Default size for each element / element size used in Typography component
+export const typographyElementDefaultSize = {
```
Is this ever used?

daksas303 replied:
This was there to specify all used sizes for tailwind, due to them being used dynamically in the past for headers. Considering we are using now specific default sizes for everything I think this can be removed.
